qcodes/tests/test_instrument.py

```diff
@@ -84,6 +84,23 @@ def test_instrument_fail(close_before_and_after):
     assert Instrument._all_instruments == {}
 
 
+def test_instrument_on_invalid_identifier(close_before_and_after):
+    # Check if error raised when invalid identifer name given
+    with pytest.raises(ValueError, match="!-name invalid instrument identifier"):
+        DummyInstrument(name="!-name")
+    assert Instrument.instances() == []
+    assert DummyInstrument.instances() == []
+    assert Instrument._all_instruments == {}
+
+    # Check if name is valid identifier when dashes '-'
+    # are converted to underscores '_'
```

Comment:
When testing code that triggers a warning its a good idea to check that the warning is actually raised. Both because it ensures that the code works correctly and because it will avoid the warning being shown when running the test suite. 
https://docs.pytest.org/en/6.2.x/warnings.html#warns shows how this can be done using pytest